### PR TITLE
Fix #4456: Pickle quote before compiling

### LINKED ISSUE #4456: Crash while running quote inside macro
```scala
import scala.quoted._
import dotty.tools.dotc.quoted.Toolbox._
object Macros {
  inline def foo(i: Int): Int = ~{
    val y = ('(i)).run
    y.toExpr
  }
}
```

```scala
import Macros._
object Test {
  def main(args: Array[String]): Unit = {
    println(foo(1))
    println(foo(1  + 3)) // Crashes the compiler
  }
}
```

### PR DESCRIPTION
**Pull request by nicolasstucki**

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/core/quoted/PickledQuotes.scala [hunk 1]
```diff
@@ -22,7 +22,26 @@ import scala.reflect.ClassTag
 object PickledQuotes {
   import tpd._
 
-  /** Pickle the quote into strings */
+  /** Pickle the tree of the a quoted.Expr */
```
Comment (biboudis) [inline compiler/src/dotty/tools/dotc/core/quoted/PickledQuotes.scala]: remove the `a` :P

File: compiler/src/dotty/tools/dotc/core/quoted/PickledQuotes.scala [hunk 2]
```diff
@@ -22,7 +22,26 @@ import scala.reflect.ClassTag
...
+          throw new scala.quoted.FreeVariableError(tree.name.toString)
+        case tree: DefTree =>
+          definedHere += tree.symbol
+          traverseChildren(tree)
+        case _ =>
+          traverseChildren(tree)
+      }
+    }.traverse(tree)
+    val pickled = pickleQuote(tree)
+    scala.runtime.quoted.Unpickler.unpickleExpr(pickled, Nil)
```
Comment (biboudis) [inline compiler/src/dotty/tools/dotc/core/quoted/PickledQuotes.scala]: The `Nil` there makes me wonder what if this fix applies to non-inline functions.

File: library/src/scala/quoted/Expr.scala [hunk 3]
```diff
@@ -37,7 +37,8 @@ object Exprs {
   }
 
   /** An Expr backed by a tree. Only the current compiler trees are allowed. */
```
Comment (biboudis) [inline library/src/scala/quoted/Expr.scala]: Clarify what we mean by "current compiler"?

**Comment by biboudis** [comment]
I tried this and it raises an exception 🤔 

```scala
import scala.quoted._
import dotty.tools.dotc.quoted.Toolbox._

object Test {
  def main(args: Array[String]): Unit = {

    def foo(i: Expr[Int] => Expr[Int]): Int = {
      val x = '(1)

      val y = xx: Expr[Int] => (i(xx).run)

      y(x)
    }

    println(foo(x => '{~x + 1}))
  }
}
```

**Comment by biboudis** [comment]
LGTM

File: compiler/src/dotty/tools/dotc/core/quoted/PickledQuotes.scala (hunk 2 above)
Comment (biboudis) [reply compiler/src/dotty/tools/dotc/core/quoted/PickledQuotes.scala]: // ignore, I have just went through the second unit test and you do what I had in mind.

**Comment by nicolasstucki** [comment]
@biboudis check if this added documentation is clear enough

**Comment by biboudis** [comment]
The description of the exception is exceptional 👍 😄

**Pull request merged**